Домашнее задание Модуль №13

Ulyanov-VN opened the pull request.

Ulyanov-VN committed:
Домашнее задание Модуль №13

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/PlayerActivity.kt:
```diff
@@ -0,0 +1,298 @@
+            .transform(RoundedCorners(resources.getDimensionPixelSize(R.dimen.corner_radius)))
+            .into(artwork)
+
+        findViewById<TextView>(R.id.trackName).text = track.trackName ?: getString(R.string.unknown_track)
```
Вызовы `findViewById` являются дорогостоящими - они долго выполняются, т.к. ищут вью в иерархии, поэтому лучше выполнять подобные вызовы 1 раз и запоминать результат в переменную, чтобы переиспользовать

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -27,6 +29,11 @@ import kotlinx.coroutines.launch
 
 class SearchActivity : BaseActivity() {
 
+    companion object {
```
`companion object` принято объявлять внизу класса согласно Kotlin Coding Convention (https://kotlinlang.org/docs/coding-conventions.html#class-layout)

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/Track.kt:
```diff
@@ -1,18 +1,76 @@
             ?.let { SimpleDateFormat("mm:ss", Locale.getDefault()).format(it) }
             ?: "--:--"
+
+    constructor(parcel: Parcel) : this(
+        parcel.readLong(),
+        parcel.readString(),
+        parcel.readString(),
+        parcel.readString(),
+        parcel.readString(),
+        parcel.readString(),
+        parcel.readString(),
+        parcel.readValue(Long::class.java.classLoader) as? Long,
+        parcel.readString(),
+        parcel.readString()
+    )
```
Для генерации реализации `Parcelable` интерфейса есть плагин `kotlin-parcelize`. Чтобы реализовать нужно сделать следующее - добавить плагин в проект, доменную модель трека наследовать от интерфейса `Parcelable` и воспользоваться аннотацией `@Parcelize`. Тут есть пример работы с плагином - https://developer.android.com/kotlin/parcelize

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/SearchViewModel.kt:
```diff
@@ -35,6 +39,28 @@ class SearchViewModel(private val repo: SearchRepository) : ViewModel() {
+    /**
+     * Восстанавливает последний поисковый запрос без выполнения нового запроса
+     */
+    fun restoreLastSearch() {
```
Восстановление в поиске насколько я вижу не совсем корректно работает - так как ты не восстанавливаешь значение в поле ввода, то отображается история поиска, а не последние результаты. Вообще, т.к. у тебя есть вьюмодель со стейтом - тебе можно не удалять значение из строки поиска и состояние будет автоматически восстанавливаться, т.к. вьюмодель его будет сохранять

ArturNurtdinov reviewed (commented):
1. Отсутствует текущее время воспроизведения под кнопкой воспроизведения, на данный момент можно отрисовать 00:00:

<img width="852" height="383" alt="image" src="https://github.com/user-attachments/assets/bbe746e2-b70a-4209-a1ff-9d52dbbf1a79" />

Макет:

<img width="829" height="466" alt="image" src="https://github.com/user-attachments/assets/f4c872ec-869b-46ee-b1d8-0c112b9dd87f" />

2. Отступ обложки, названия трека, названия исполнителя и боковых кнопок от края экрана должно быть 24dp, сейчас - 40dp (т.к. родительский Constraint имеет padding=16dp и при этом элементы имеют дополнительный отступ в 24dp (итого 24+16=40 вместо ожидаемых 24):

<img width="567" height="830" alt="image" src="https://github.com/user-attachments/assets/2a2332ca-a361-4e82-b52b-3ce09cfb40d8" />

Макет:

<img width="554" height="818" alt="image" src="https://github.com/user-attachments/assets/0e91a86d-1a04-43e5-b182-b5d7734ab63a" />

Ulyanov-VN committed:
Домашнее задание Модуль №13 (устранение замечаний)

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/PlayerActivity.kt:
```diff
@@ -33,6 +33,8 @@ class PlayerActivity : AppCompatActivity() {
         window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN)
         setContentView(R.layout.activity_player)
 
+        findViewById<TextView>(R.id.trackDuration).text = "00:00"
```
Лучше не хардкодить этот текст, так как форматирование может легко поменяться и потом придётся по всему проекту искать эти захардкоженные строчки. Можно либо вынести в strings.xml, чтобы это было только в одном месте, либо использовать ту же функцию форматирования через `SimpleDateFormat`, передавая туда значение 0, чтобы поведение было консистентным

ArturNurtdinov commented on app/src/main/res/layout/activity_player.xml:
```diff
@@ -5,14 +5,15 @@
         android:layout_width="wrap_content"
         android:layout_height="wrap_content"
+        android:layout_marginStart="24dp"
+        android:layout_marginTop="24dp"
```
У кнопки был отступ в 16dp за счёт `padding`, ты убрал `padding` и увеличил отступ до 24, что уже не соответствует макетам.

ArturNurtdinov commented on app/src/main/res/layout/activity_player.xml:
```diff
@@ -25,14 +26,15 @@
         android:layout_height="0dp"
-        android:scaleType="centerCrop"
         android:layout_marginTop="32dp"
+        android:layout_marginStart="24dp"
+        android:layout_marginEnd="24dp"
```
Также у текста снизу был отступ в 16dp, что было корректно, сейчас - 24

Ulyanov-VN committed:
Домашнее задание Модуль №13 (повтороное устранение замечаний)

1. Исправил "Отступ у кнопки "Назад" сверху экрана и слева должен быть 16dp (сейчас 24)": 1. Обратился к наставнику за советом он мне предложил сделать margin = "4dp", а padding="12dp".  2. добавил android:fitsSystemWindows="true" чтобы стрелка не налезала на тулбар. 3. Добавил инокнку 24x24 и сделал корректировку с учетом этой иконки

2. Исправил "Горизонтальный отступ у текстовых полей под временем воспроизведения должен быть 16dp, сейчас - 24dp " строке длительность добавлен android:layout_marginStart="16dp" и android:layout_marginEnd="16dp", а все остальные строки наследуют этот параметр.

Все остальные твои коментарии я сохранил и буду разбираться. Спасибо большое.
